validator/src/aggregation/tendermint/protocol.rs

```diff
@@ -37,14 +35,60 @@ impl TendermintAggregationProtocol {
             ReplaceStore::<TendermintIdentifier, Self>::new(Arc::clone(&partitioner)),
         ));
 
+        // This will be moved into the closure.
+        let cloned_validators = Arc::clone(&validators);
+
         let evaluator = Arc::new(WeightedVote::new(
             Arc::clone(&store),
-            validators.clone(),
+            Arc::clone(&validators),
             Arc::clone(&partitioner),
-            threshold,
+            move |message: &TendermintContribution| {
+                // Conditions for final contributions:
+                // * Any proposal has 2f+1
+                // (* Nil has f+1) would be implicit in the next condition
+                // * The biggest proposal (NOT Nil) combined with the non cast votes is less than 2f+1
+
+                let mut uncast_votes = policy::Policy::SLOTS as usize;
+                let mut biggest_weight = 0;
+
+                // Check every proposal present, including None
+                for (hash, signature) in message.contributions.iter() {
+                    let weight = cloned_validators
+                        .signers_weight(&signature.signers)
+                        .unwrap_or(0);
+                    // Any option which has at least 2f+1 votes make this contribution final.
+                    if weight >= policy::Policy::TWO_F_PLUS_ONE as usize {
+                        return true;
+                    }
+                    // If the proposal does not have 2f+1 check if it has the new biggest_weight
+                    if hash.is_some() && weight > biggest_weight {
+                        biggest_weight = weight;
+                    }
+                    // The cast votes need to be removed from the uncast votes
+                    uncast_votes -= weight;
```

Comment:
Couldn't we better use `saturating_sub`?

Reply:
I use `saturating_sub` here now. Note that if this would ever saturate the contribution would also be invalid as a whole as then there is more votes present as there is votes to begin with. A different place would have already filtered this out. 
It is still way better to use the `saturating_sub` here.